src/Mod/Part/App/TopoShapeExpansion.cpp

```diff
@@ -4237,9 +4237,33 @@ TopoShape& TopoShape::makeElementPrismUntil(const TopoShape& _base,
         // When using the face with BRepFeat_MakePrism::Perform(const TopoDS_Shape& Until)
         // then the algorithm expects that the 'NaturalRestriction' flag is set in order
         // to work as expected.
-        BRep_Builder builder;
-        _uptoface = _uptoface.makeElementCopy();
-        builder.NaturalRestriction(TopoDS::Face(_uptoface.getShape()), Standard_True);
+
+        // skip concave face
+        bool isConcave = false;
+        TopoDS_Face theFace = TopoDS::Face(_uptoface.getShape());
+        Handle(Geom_Surface) surf = BRep_Tool::Surface(theFace);
+        GeomAdaptor_Surface adapt(surf);
+        if(adapt.GetType() != GeomAbs_Plane){
+            // get parameters bounds
+            Standard_Real uMin, uMax, vMin, vMax;
+            surf->Bounds(uMin, uMax, vMin, vMax);
+            // compute normals at the middle
+            // TODO? find intersection point in UpToFace
+            gp_Pnt pt;
+            gp_Vec dU, dV;
+            surf->D1((uMin+uMax)/2, (vMin+vMax)/2, pt, dU, dV);
+            // check normals orientation
+            gp_Dir dirdU(dU);
+            isConcave = dirdU.Angle(direction) < M_PI_2;
+            gp_Dir dirdV(dV);
+            isConcave = isConcave || (dirdV.Angle(direction) < M_PI_2);
+        }
+
+        if (! isConcave){
+            BRep_Builder builder;
+            _uptoface = _uptoface.makeElementCopy();
+            builder.NaturalRestriction(TopoDS::Face(_uptoface.getShape()), Standard_True);
+        }
```

Comment:
@realthunder what do you think about this? It works for [my test case](https://github.com/FreeCAD/FreeCAD/issues/14663#issuecomment-2166651538) but you could imagine a failure scenario.